src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs

```diff
@@ -1636,14 +1636,25 @@ internal static CollectionExpressionTypeKind GetCollectionExpressionTypeKind(CSh
                     return CollectionExpressionTypeKind.Array;
                 }
             }
-            else if (isSpanType(compilation, destination, WellKnownType.System_Span_T, out elementType))
+            else if (destination is NamedTypeSymbol namedType
+                && namedType.OriginalDefinition.Equals(compilation.GetWellKnownType(WellKnownType.System_Collections_Immutable_ImmutableArray_T), TypeCompareKind.ConsiderEverything)
+                && compilation.GetWellKnownTypeMember(WellKnownMember.System_Runtime_InteropServices_ImmutableCollectionsMarshal__AsImmutableArray_T) is not null)
```

Comment:
>&& compilation.GetWellKnownTypeMember(WellKnownMember.System_Runtime_InteropServices_ImmutableCollectionsMarshal__AsImmutableArray_T) is not null [](http://example.com/codeflow?start=16&length=145)

It seems a little surprising we're setting the `CollectionExpressionTypeKind` based on whether the optimization is available. Could we simply use the existing `CollectionBuilder` code path in binding and then check for `ImmutableCollelctionsMarshal.AsImmutableArray<T>()` in lowering in `LocalRewriter.VisitCollectionBuilderCollectionExpression()`?

Reply:
Basically, I can see how having to check the same thing in both conversions checks and in lowering is a little cumbersome, but I also think that it's reasonable to check for existence of certain symbols when deciding which `CollectionExpressionTypeKind` to use. Since we do this same thing with the `[CollectionBuilder]` attribute and so on.